torch/quantization/fx/quantization_patterns.py

```diff
@@ -143,32 +144,44 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
             else:
                 return quantizer.quantized_graph.node_copy(node, load_arg(quantized=False))
 
-        if self.num_node_args == 1:
-            # add/mul scalar
-            if isinstance(self.bop_node.args[0], Node):
-                quantized_index = 0
-            else:
-                quantized_index = 1
-
-            return quantizer.quantized_graph.create_node(
-                'call_function', self.qop,
-                load_arg(quantized=[quantized_index])(self.bop_node.args), self.bop_node.kwargs)
-        else:
-            activation_post_process = quantizer.activation_post_process_map[node.name]
-            scale, zero_point = activation_post_process.calculate_qparams()
-            scale = float(scale)
-            zero_point = int(zero_point)
-            scale_arg, zero_point_arg = create_qparam_nodes(quantizer, node.name, scale, zero_point)
+        if dtypes in [(torch.quint8, torch.qint8, None)]:
+            if self.num_node_args == 1:
+                # add/mul scalar
+                if isinstance(self.bop_node.args[0], Node):
+                    quantized_index = 0
+                else:
+                    quantized_index = 1
 
-            if self.relu_node is not None:
-                op = torch.ops.quantized.add_relu
+                return quantizer.quantized_graph.create_node(
+                    'call_function', self.qop,
+                    load_arg(quantized=[quantized_index])(self.bop_node.args), self.bop_node.kwargs)
             else:
-                op = torch.ops.quantized.add
-            kwargs = {**self.bop_node.kwargs}
-            add_args = (*load_arg(quantized=True)(self.bop_node.args), scale_arg, zero_point_arg)
-            op = quantizer.quantized_graph.create_node(
-                'call_function', self.qop, add_args, kwargs)
-            return op
+                activation_post_process = quantizer.activation_post_process_map[node.name]
+                scale, zero_point = activation_post_process.calculate_qparams()
+                scale = float(scale)
+                zero_point = int(zero_point)
+                scale_arg, zero_point_arg = create_qparam_nodes(quantizer, node.name, scale, zero_point)
+
+                if self.relu_node is not None:
+                    op = torch.ops.quantized.add_relu
+                else:
+                    op = torch.ops.quantized.add
+                kwargs = {**self.bop_node.kwargs}
+                add_args = (*load_arg(quantized=True)(self.bop_node.args), scale_arg, zero_point_arg)
+                op = quantizer.quantized_graph.create_node(
+                    'call_function', self.qop, add_args, kwargs)
+                return op
+        elif dtypes in [(torch.float16, torch.float16, None)]:
```

Comment:
just so I understand correctly...
1. this code copies the add or mul and relu
2. the code in the quantizer object will insert the fake quantize to fp16 logic, using `x.to(torch.half)`

?

Reply:
1. yes that's correct, in fp16 reference pattern we just copy add/mul op, the transformation we do happens in load_arg(quantized=False), which dequantizes the fp16 inputs.
2. not sure what you mean, the magic happens when people call quantize_node (which inserts x.to(torch.half)) and when we call load_arg(quantized=False), which inserts x = x.dequantize() (we can dequantize a float16 Tensor to float32 by calling to now: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/QTensor.cpp#L46)